### ISSUE TITLE
add pipe functionality to nyc for instrumentation

### ISSUE THREAD

**Issue by bcoe** [istanbuljs/nyc#516]
See [conversation with @addaleax](https://github.com/nodejs/node/pull/10856#issuecomment-277821130):

__might help quite a bit if we could reduce the startup overhead of nyc instrument – our tooling would work more nicely if we could use nyc instrument as a pipe-in, pipe-out command for each source file individually, but right now that’s not really practical. I’ve played around a bit but didn’t get too far (it still always had upwards of 0.7 seconds/run overhead).__

## Expected Behavior

This is essentially two feature requests:

1. it would be nice to be able to pipe arbitrary source code through `nyc` and get instrumented code out the other side (I'm correct in this summarization @addaleax?). Something like:
  * `cat source.js | nyc`
2. it would be nice if we could reduce nyc's startup time in general (this would probably be work separate from the piping work, and we will likely want to break this into another ticket).

CC: @CurryKitten

**Comment by addaleax** [comment]
The 1. part is actually not that important for the Node core stuff (although I think it’s something `nyc` should support, and I don’t think it would be too hard to implement).

The second part is probably trickier ;)

**Comment by bcoe** [comment]
@addaleax we do a few heavy handed things during yargs' startup, e.g., loading configuration from package.json using `find-up`; it might be the case that we could shuffle this logic around a bit and significantly improve boot time.

**Comment by bcoe** [comment]
as I understand it, the issue is that you're spending `600ms` before any instrumentation even begins to take place?

**Comment by addaleax** [comment]
> as I understand it, the issue is that you're spending `600ms` before any instrumentation even begins to take place?

Yup, exactly. Take that times a hundred source files and you’ll have a minute. ;)

We’ll be able to around that in some way, but if we *can* get nyc’s startup time down and bring it into the usable range that would make it usable for per-file work, that would be a huge win (for everyone, I guess :smile:)
